### PR TITLE
Change CLangAnalyzer to stream issues

### LINKED ISSUE #1471: Display CFamily analysis issues incrementally
### Description

Streaming CFamily issues from `subprocess.exe` allows us to display new issues as they are reported, rather than waiting for all of the issues to reported before updating the UI.

There are many different user experiences we could build on top of the new streaming mechanism. 
The proposed workflow described below is the simplest from an implementation point of view. We will implement this solution and evaluate whether the resulting UX is acceptable.

#### Status reporting
There should be a visible indication that analysis has started, is in progress, and has completed (either successfully or unsuccessfully).
In this sprint we are not planning on any custom UI feedback mechanisms. We will use only those provided out of the box by Visual Studio. The VS guidance on progress indicators is [here](https://docs.microsoft.com/en-us/visualstudio/extensibility/ux-guidelines/notifications-and-progress-for-visual-studio?view=vs-2019#BKMK_ProgressIndicators).

#### Proposed initial workflow
1. Trigger analysis for a file, streaming the results from `subprocess.exe`
2. When the first issue arrives, clear the existing issues for the file
3. Create a new snapshot containing the new issue
4. Update the error list
5. Repeat steps (3) and (4) as further issues are reported; each time, the new snapshot will contain one more issue than the previous snapshot.

TODO: decide when to update the tags in the editor. This could be done at the same time the error list is updated, or at the end.

#### Special cases to consider
* a new analysis is triggered before the previous analysis has completed: check the in-process analysis is cancelled
* ?time limit for a single file analysis? Infinite? One hour? One minute?
* highlighting the correct line of text in the editor: e.g. if an issue is reported after twenty seconds, there is a risk that the user will have edited the file in the meantime so the issue location will have changed. This is not a new issue, although it is more likely to occur if we increase the time an analysis can run before it is cancelled. CHECK: do we currently handle this scenario correctly?

#### UX Risks
Removing all of the old issues at once might be visually distracting.
Pushing an update to the error list/editor as each new issue arrives might negatively impact the responsiveness of the UI. It might also be visually distracting to have the UI update repeatedly as new issues arrive (e.g. if the error list re-ordered each time).

### PR DESCRIPTION
**Pull request by duncanp-sonar**
Change CLangAnalyzer to stream issues

* removed message about suppressed issues

Fixes #1471

### CODE REVIEW

File: src/Integration.Vsix/CFamily/CLangAnalyzer.cs [hunk 1]
```diff
@@ -99,33 +99,35 @@ private async Task TriggerAnalysisAsync(Request request, IIssueConsumer consumer
...
-
-            var issues = messages
-                .Where(m => IsIssueForActiveRule(m, request.RulesConfiguration))
-                .Select(m => CFamilyHelper.ToSonarLintIssue(m, request.CFamilyLanguage, request.RulesConfiguration))
-                .ToList();
-
             telemetryManager.LanguageAnalyzed(request.CFamilyLanguage); // different keys for C and C++
 
-            logger.WriteLine($"Found {issues.Count} issue(s)");
+            logger.WriteLine($"Found {issueCount} issue(s)");
```
Comment (rita-g-sonarsource) [inline src/Integration.Vsix/CFamily/CLangAnalyzer.cs:112]: I think it's worth adding the fileName into the message. There could be multiple analysis running at the same time.

File: src/Integration.Vsix/SonarLintTagger/TextBufferIssueTracker.cs [hunk 2]
```diff
@@ -271,9 +271,6 @@ private IEnumerable<IssueMarker> RemoveSuppressedIssues(IEnumerable<IssueMarker>
             var filteredIssues = issuesFilter.Filter(filterableIssues);
             Debug.Assert(filteredIssues.All(x => x is IssueMarker), "Not expecting the issue filter to change the list item type");
 
-            var suppressedCount = filterableIssues.Count() - filteredIssues.Count();
```
Comment (rita-g-sonarsource) [inline src/Integration.Vsix/SonarLintTagger/TextBufferIssueTracker.cs:274]: If we have an easy/elegant way to identify the issue, we could output a generic "Issue X has been suppressed"

Comment (duncanp-sonar) [reply src/Integration.Vsix/SonarLintTagger/TextBufferIssueTracker.cs:274]: Even if there was, we'd see the message multiple times, in the same way we say the suppressed issue count message multiple times.

Comment (rita-g-sonarsource) [reply src/Integration.Vsix/SonarLintTagger/TextBufferIssueTracker.cs:274]: Ah right.

File: src/Integration.Vsix/CFamily/CLangAnalyzer.cs [hunk 3]
```diff
@@ -89,7 +89,13 @@ public bool IsAnalysisSupported(IEnumerable<AnalysisLanguage> languages)
             TriggerAnalysisAsync(request, consumer, cancellationToken)
                 .Forget(); // fire and forget
 
-        private async Task TriggerAnalysisAsync(Request request, IIssueConsumer consumer, CancellationToken cancellationToken)
+        protected /* for testing */ virtual void CallSubProcess(Action<Message> handleMessage, Request request, IAnalysisStatusNotifier analysisStatusNotifier, ISonarLintSettings settings,
```
Comment (duncanp-sonar) [inline src/Integration.Vsix/CFamily/CLangAnalyzer.cs:92]: I originally define a delegate, but it looked weird to have a mix of delegates and `virtual override` points in the class (the existing two virtual methods are defined immediately above this one).
I also looked at creating an interface, but that involved more change. This was the minimum change I could make to the product code to make the behaviour testable.

File: src/Integration.Vsix.UnitTests/CFamily/CLangAnalyzerTests.cs [hunk 4]
```diff
@@ -137,6 +241,51 @@ protected override void TriggerAnalysis(Request request, IIssueConsumer consumer
...
+                    base.CallSubProcess(handleMessage, request, analysisStatusNotifier, settings, logger, cancellationToken);
+                }
+                else
+                {
+                    onCallSubProcess(handleMessage, request, analysisStatusNotifier, settings, logger, cancellationToken);
+                }
+            }
+        }
+
+        private class SubProcessSimulator
```
Comment (duncanp-sonar) [inline src/Integration.Vsix.UnitTests/CFamily/CLangAnalyzerTests.cs:258]: This class is ugly but necessary of the way the CLangAnalyzer expects the stream handler to not return until all of the issues have been received.

File: src/Integration.Vsix.UnitTests/CFamily/CLangAnalyzerTests.cs [hunk 5]
```diff
@@ -100,6 +104,99 @@ public void ExecuteAnalysis_ValidProjectItem_RequestCanBeCreated_AnalysisIsTrigg
             testSubject.TriggerAnalysisCallCount.Should().Be(1);
         }
 
+        [TestMethod]
+        public void TriggerAnalysisAsync_StreamsIssuesFromSubProcessToConsumer()
```
Comment (duncanp-sonar) [inline src/Integration.Vsix.UnitTests/CFamily/CLangAnalyzerTests.cs:108]: The new tests have to call the analyzer on a different thread from the test thread, and then selectively block/wait as the messages are processed.
As far as I can tell, the tests are reliable (i.e. repeatably run correctly without flakiness) and also clean up properly in the event of errors.

File: src/Integration.Vsix/CFamily/CLangAnalyzer.cs [hunk 6]
```diff
@@ -99,33 +105,38 @@ private async Task TriggerAnalysisAsync(Request request, IIssueConsumer consumer
...
+                logger.WriteLine($"Found {issueCount} issue(s) for {request.File}");
+            }
+        }
+
+        private void HandleMessage(Message message, Request request, IIssueConsumer consumer, ref int issueCount)
+        {
+            Debug.Assert(message.Filename == request.File, $"Issue for unexpected file returned: {message.Filename}");
+            if (!IsIssueForActiveRule(message, request.RulesConfiguration))
+            {
+                return;
```
Comment (duncanp-sonar) [inline src/Integration.Vsix/CFamily/CLangAnalyzer.cs:129]: According to SonarCloud there are 3uncovered lines of code out of 7, giving 57% coverage. This is one of the uncovered lines, which I can add a test for quite easily.

The other two are for the new virtual method declaration and the one-line passthough call to CFamilyHelper. Testing that would require an integration tests, which I'm not proposing to do at this point.

Comment (rita-g-sonarsource) [reply src/Integration.Vsix/CFamily/CLangAnalyzer.cs:129]: I agree. It's worth testing the non-active rule logic.

File: src/Integration.Vsix.UnitTests/CFamily/CLangAnalyzerTests.cs [hunk 7]
```diff
@@ -137,6 +241,51 @@ protected override void TriggerAnalysis(Request request, IIssueConsumer consumer
...
+                // Store the callback passed to us from the CLangAnalyzer
+                handleMessageCallback = handleMessage;
+
+                // Tell the calling test we're ready and the test can continue
+                callbackFromCLangReceived.Set();
+
+                // Block until the test tells us we can finish
+                noMoreIssues.WaitOne();
+            }
+
```
Comment (rita-g-sonarsource) [inline src/Integration.Vsix.UnitTests/CFamily/CLangAnalyzerTests.cs:280]: I feel stupid. Why do we need this?
Why can't we just use `noMoreIssues`?

**Comment by rita-g-sonarsource** [comment]
LGTM

**Pull request merged**